src/mds/MDCache.cc

```diff
@@ -8651,7 +8651,7 @@ int MDCache::path_traverse(MDRequestRef& mdr, MDSContextFactory& cf,
     ceph_abort();  // i shouldn't get here
   }
 
-  if (want_auth && !want_dentry) {
```

Comment:
IMO this is incorrect. As we discussed in the mail once here current MDS should be the auth MDS for the tail dentry in the openc path.

I think you should trigger the reintegrate the stray dentry instead ?

Reply:
> IMO this is incorrect. As we discussed in the mail once here current MDS should be the auth MDS for the tail dentry in the openc path.
> 
> I think you should trigger the reintegrate the stray dentry instead ?

https://github.com/ceph/ceph/pull/49553
https://tracker.ceph.com/issues/55332
This is the fix about unlink.
In production environment, from the log, the crash is indeed related to unlink. 
LOG:(handle_client_openc -->rdlock_path_xlock_dentry --> path_traverse)
  mds.2.cache linking in remote in [inode 0x30004be7edb [...2,head] ~mds4/stray1/30004be7edb rep@4.1 v172464532 snaprealm=0x55eb2dc0a500 s=2203120 n(v0 rc2022-12-31T22:44:39.135649+0800 b2203120 1=1+0)               (iversion lock) 0x55eb2ce3bc00]

Reply:
> IMO this is incorrect. As we discussed in the mail once here current MDS should be the auth MDS for the tail dentry in the openc path.
> 
> I think you should trigger the reintegrate the stray dentry instead ?

@lxbsz Could you please share the discussion here ?

Reply:
@kotreshhr Sorry, that's in Chinese. The `_openc()`  failed when it detected the `CInode` is linked in the stray. That means the `CInode` was deleted and assigned to a new dentry, then we should eval to reintegrate the dentry before continuing further.

@orange30 Could you raise one tracker for this and paste the detail logs in the tracker as you sent to me ? Thanks.

Reply:
This change does not explain how the crash is fixed. I suggest creating a tracker ticket with the mds logs from the crash.

Reply:
FWIW, when linking in the remote inode, there is a check if the inode can be reintegrated.

Reply:
> I updated the tracker, I am sorry for upload it late. I think this commit + #46331 can fix the crash problem. @vshankar @lxbsz

@orange30 Could you explain how this commit fixes the crash? Sorry, but I do not see how.

Reply:
> FWIW, when linking in the remote inode, there is a check if the inode can be reintegrated.

Yeah, but apparently this case doesn't trigger that.

Reply:
> > I updated the tracker, I am sorry for upload it late. I think this commit + #46331 can fix the crash problem. @vshankar @lxbsz
> 
> @orange30 Could you explain how this commit fixes the crash? Sorry, but I do not see how.

Let's analyze this log：
LOG:(handle_client_openc -->rdlock_path_xlock_dentry --> path_traverse)
mds.2.cache linking in remote in [inode 0x30004be7edb [...2,head] ~mds4/stray1/30004be7edb rep@4.1 v172464532 snaprealm=0x55eb2dc0a500 s=2203120 n(v0 rc2022-12-31T22:44:39.135649+0800 b2203120 1=1+0) (iversion lock) 0x55eb2ce3bc00]

1, If we use this commit, the request will be forward to the auth mds, for this log case: the openc request will be forward to mds4.
2, On mds4, in handle_client_openc, because the dn has STATE_UNLINKING flag, the request will be added to wait list：
      if (is_unlink_pending(dn)) {
        wait_for_pending_unlink(dn, mdr);
        return;
      }

Reply:
> I updated the tracker, I am sorry for upload it late. I think this commit + #46331 can fix the crash problem. @vshankar @lxbsz

The PR #46331 should be a different issue, which is create the same dentry, the same dentry name, instead of the reusing the same inode#.

Reply:
> > FWIW, when linking in the remote inode, there is a check if the inode can be reintegrated.
> 
> Yeah, but apparently this case doesn't trigger that.

Not really. Commit 83f9a49dfddcba06ab3e0cd9b860523fafad4605 add a reintegration check when loading a dentry which isn't available in octopus release.

Reply:
Okay, I think you are right.

Reply:
> 

UPDATE:
On mds4:
(1) If the inode is unlinked by the same dentry with the dentry in the openc, the request will be added to wait list(the dn has STATE_UNLINKING flag).
(2) If the inode is unlinked by the other hard link, because the dentry in the openc donot has STATE_UNLINKING flag, so the request will not be added to dn wait list. Further processing is required here.

HERE:
If we can get dn successfully from rdlock_path_xlock_dentry in handle_client_openc, and if dnl->is_null() is false, it mean this file has been created before (even if the inode is under ~mds4/stray1/).

So the question can be translated:
A file has hardlink1 and handlink2. When the unlink from hardlink1 is not finished, if we want open the file from hardlink2, what should we do?

Reply:
> So the question can be translated: A file has hardlink1 and handlink2. When the unlink from hardlink1 is not finished, if we want open the file from hardlink2, what should we do?

I think the crash you are seeing in most likely fixed in recent ceph versions as mentioned in https://github.com/ceph/ceph/pull/49652#discussion_r1065359835. Would you mind trying p/q release is possible?

Reply:
I will have a look at the commit #49652.

Another question for help:
If the mds can received openc request, does this mean: on the client the dentry name in the openc request  donot existed now ? (not created or has been deleted)
If this is right, this mean: the handle_client_openc->handle_client_openc case should not be the case: "A file has hardlink1 and handlink2. .... what should we do?"

Reply:
> I will have a look at the commit [#49652 (comment)](https://github.com/ceph/ceph/pull/49652#discussion_r1065359835).
> 
> Another question for help: If the mds can received openc request, does this mean: on the client the dentry name in the openc request donot existed now ? (not created or has been deleted)

Client can return -EEXIST without sending a request to the MDS when O_CREAT and O_EXCL are provided after path walk.

> If this is right, this mean: the handle_client_openc->handle_client_open case should not be the case: "A file has hardlink1 and handlink2. .... what should we do?"

In the crash you are seeing, O_EXCL is unset, so the client makes a request to the MDS.

Reply:
"mds.2.server open flags = 577"  -- 0x241
#define CEPH_O_EXCL            00000200 
The flags CEPH_O_EXCL is in the flags.
CEPH_O_EXCL looks like not coming from O_EXCL directly.

Reply:
The MDS logs your shared in the tracker does not have O_EXCL set in the open call flags.

Reply:
As so far, from the crash log we have mentioned three potential problems.

Let's discuss it on the contidion: a file has hardlink1 and handlink2, and the Inode's auth mds is mds4.

The first potential problem:
    If we open the handlink2 on mds2 with flag CEPH_O_TRUNC, the mds will crash, and the crash stack is the same with this crash log. 
    (1) this problem is nothing to do with unlink.
    (2) the latest ceph version also have this problem.
    This commit can fix this problem.

The second potential problem:
    If we delete hardlink1 on mds4 and before this unlink is finished, we create hardlink1 again.
	This problem has been fixed by #46331 .

The third protential problem:
    If we delete hardlink1 on mds4 and before this unlink is finished, we open hardlink2 on mds2.
    This problem maybe has been fixed by https://github.com/ceph/ceph/commit/83f9a49dfddcba06ab3e0cd9b860523fafad4605 , I will have a look it.	

Do you agree with this? @vshankar @lxbsz

Reply:
> As so far, from the crash log we have mentioned three potential problems.
> 
> Let's discuss it on the contidion: a file has hardlink1 and handlink2, and the Inode's auth mds is mds4.
> 
> The first potential problem: If we open the handlink2 on mds2 with flag CEPH_O_TRUNC, the mds will crash, and the crash stack is the same with this crash log. (1) this problem is nothing to do with unlink. (2) the latest ceph version also have this problem.

I ran the test for https://tracker.ceph.com/issues/58340 and checked the logs, the above case you mentioned should exist, such as:

If the `testdir/`, which the auth MDS is `mds0`, has one dentry naming as `testdn`, but the `testdn` is just a hardlink, which the auth MDS is `mds1` instead.

That means if you try to `open & trunc` the `testdn` in `mds0` and then in the `_openc()->rdlock_path_xlock_dentry()->path_traverse()` it will return the `testdn` without forwarding to `mds1`.

And then later in `_openc()->_open()->rdlock_path_pin_ref()` it will skip run the `path_traverse()` because the `mdr->locking_state & MutationImpl::PATH_LOCKED` is true, which was set in the previous `path_traverse()`.

So in `mds0` the `testdn`'s inode it's auth is in `mds1`.
 
I think your analysis is correct, but your fix is incorrect and I am afraid it will cause the request will ping-pong between the `mds0` and `mds1 inifinitely, because after you forwarding the request to `mds1` then the `testdir/` in `mds1` won't be the auth MDS and then it will be forwarded back in `_openc()->rdlock_path_xlock_dentry()->path_traverse()` ?

Reply:
In mds1, for testdir/, beacuse it is just the middle path, I think path_traverse() will not forward back to msd0.

Reply:
> As so far, from the crash log we have mentioned three potential problems.
> 
> Let's discuss it on the contidion: a file has hardlink1 and handlink2, and the Inode's auth mds is mds4.
> 
> The first potential problem: If we open the handlink2 on mds2 with flag CEPH_O_TRUNC, the mds will crash, and the crash stack is the same with this crash log. (1) this problem is nothing to do with unlink. (2) the latest ceph version also have this problem. This commit can fix this problem.

Sounds like this is straightforward to reproduce. Would you mind including a test case for this?

> 
> The second potential problem: If we delete hardlink1 on mds4 and before this unlink is finished, we create hardlink1 again. This problem has been fixed by #46331 .

Right.

> 
> The third protential problem: If we delete hardlink1 on mds4 and before this unlink is finished, we open hardlink2 on mds2. This problem maybe has been fixed by [83f9a49](https://github.com/ceph/ceph/commit/83f9a49dfddcba06ab3e0cd9b860523fafad4605) , I will have a look it.

Right. It does not harm to add tests for this scenario. I think it might be a bit hard to reproduce since it depends on timings of certain events, however you could do the required operation in a loop for (say) 10k iterations. You could use directory pinning to pin directories to different MDSs.

Reply:
> In mds1, for testdir/, beacuse it is just the middle path, I think path_traverse() will not forward back to msd0.

Won't it  be forwarded in case:

```
 8417     if (want_auth && want_dentry && depth == path.depth() - 1) {
 8418       if (curdir->is_ambiguous_auth()) {  
 8419         dout(10) << "waiting for single auth on " << *curdir << dendl;
 8420         curdir->add_waiter(CInode::WAIT_SINGLEAUTH, cf.build());
 8421         return 1;
 8422       }
 8423       if (!curdir->is_auth()) {
 8424         dout(10) << "fw to auth for " << *curdir << dendl;
 8425         request_forward(mdr, curdir->authority().first);
 8426         return 2;
 8427       }
 8428     }
```

The `curdir` will be the `testdir`, which is the `testdn`'s parent directory ?

Reply:
I think currently it just assume the auth MDS for both `testdir/` and `testdn`  will be the same.

Reply:
> > As so far, from the crash log we have mentioned three potential problems.
> > Let's discuss it on the contidion: a file has hardlink1 and handlink2, and the Inode's auth mds is mds4.
> > The first potential problem: If we open the handlink2 on mds2 with flag CEPH_O_TRUNC, the mds will crash, and the crash stack is the same with this crash log. (1) this problem is nothing to do with unlink. (2) the latest ceph version also have this problem.
> 
> I ran the test for https://tracker.ceph.com/issues/58340 and checked the logs, the above case you mentioned should exist, such as:
> 
> If the `testdir/`, which the auth MDS is `mds0`, has one dentry naming as `testdn`, but the `testdn` is just a hardlink, which the auth MDS is `mds1` instead.
> 
> That means if you try to `open & trunc` the `testdn` in `mds0` and then in the `_openc()->rdlock_path_xlock_dentry()->path_traverse()` it will return the `testdn` without forwarding to `mds1`.

Does this looks like a bug in MDCache::path_traverse? rdlock_path_xlock_dentry does set MDS_TRAVERSE_WANT_AUTH, so ideally it should forward the request to the auth MDS. However, there is special handling for remote inodes where path_traverse would link the remote inode in the dentry (call to CDentry::link_remote). Maybe the assert check can be relaxed? I'm not sure atm - I need to do more code reading and testing.

Reply:
> Does this looks like a bug in MDCache::path_traverse? rdlock_path_xlock_dentry does set MDS_TRAVERSE_WANT_AUTH, so ideally it should forward the request to the auth MDS. 

If my understanding is correct and if the `MDS_TRAVERSE_WANT_DENTRY` flag is also set it will forward the request to auth MDS only for the tail dentry's parent directory. Not for the tail dentry.

So it could be that after `path_traverse()` the tail dentry returned could be non-authed in current MDS.

While the `rdlock_path_pin_ref()` won't set the `MDS_TRAVERSE_WANT_DENTRY` flag, so the `path_traverse()` will forward for the tail dentry if currently it's not auth, but in the `_openc()->_open()` it doesn't trigger the `path_traverse()`.

> However, there is special handling for remote inodes where path_traverse would link the remote inode in the dentry (call to CDentry::link_remote). Maybe the assert check can be relaxed? I'm not sure atm - I need to do more code reading and testing.

It seems not working. The TRUNC seems must be done on the auth MDS.

Reply:
> 



> > In mds1, for testdir/, beacuse it is just the middle path, I think path_traverse() will not forward back to msd0.
> 
> Won't it be forwarded in case:
> 
> ```
>  8417     if (want_auth && want_dentry && depth == path.depth() - 1) {
>  8418       if (curdir->is_ambiguous_auth()) {  
>  8419         dout(10) << "waiting for single auth on " << *curdir << dendl;
>  8420         curdir->add_waiter(CInode::WAIT_SINGLEAUTH, cf.build());
>  8421         return 1;
>  8422       }
>  8423       if (!curdir->is_auth()) {
>  8424         dout(10) << "fw to auth for " << *curdir << dendl;
>  8425         request_forward(mdr, curdir->authority().first);
>  8426         return 2;
>  8427       }
>  8428     }
> ```
> 
> The `curdir` will be the `testdir`, which is the `testdn`'s parent directory ?

The curdir is one of testdir's dirfrags (whose type is class CDir{} and it is defined in class CInode: mempool::mds_co::compact_map<frag_t,CDir*> dirfrags; // cached dir fragments under this Inode ).
For example, there is 100000 files in 'testdir' and it has three cdir: cdir1,cdir2 and cdir3, and we assume that 'testdn' is under cdir3.
So I think the problem can be translated to be: Is cidr3's auth mds the same with testdn's auth mds.
Now i think they are the same.

Reply:
> The curdir is one of testdir's dirfrags (whose type is class CDir{} and it is defined in class CInode: mempool::mds_co::compact_map<frag_t,CDir*> dirfrags; // cached dir fragments under this Inode ). For example, there is 100000 files in 'testdir' and it has three cdir: cdir1,cdir2 and cdir3, and we assume that 'testdn' is under cdir3. 

IMO you can just ignore the dirfrags, just assume it's directory here.

> So I think the problem can be translated to be: Is cidr3's auth mds the same with testdn's auth mds. Now i think they are the same.

It seems not. If so this crash shouldn't be triggered.

Reply:
A normal file's(not hardlink) inode's auth mds is the same with curdir.
If this is right, we can make some special process for hardlink file.

Reply:
(1)curdir(cidr3)'s auth mds
(2)testdn's dentry's auth mds
(3)testdn's inode's auth mds

I think curdir(cidr3)'s auth mds is the same with testdn's dentry's auth mds.
1, Is this right?
2, If a file is hardlink, in its dentry, can we check it is a hardlink easily ?

Reply:
The `_openc()` is special and maybe you can just drop the locks and do it again in the `_open()` as needed:

```
diff --git a/src/mds/Server.cc b/src/mds/Server.cc
index 70644db563f..a477b8244b9 100644
--- a/src/mds/Server.cc
+++ b/src/mds/Server.cc
@@ -3564,8 +3564,13 @@ CInode* Server::rdlock_path_pin_ref(MDRequestRef& mdr,
   const filepath& refpath = mdr->get_filepath();
   dout(10) << "rdlock_path_pin_ref " << *mdr << " " << refpath << dendl;
 
-  if (mdr->locking_state & MutationImpl::PATH_LOCKED)
+  if (mdr->locking_state & MutationImpl::PATH_LOCKED) {
+    if (want_auth && !mdr->in[0]->is_auth()) {
+      request_forward(mdr, mdr->in[0]->authority().first);
+      return nullptr;
+    }
     return mdr->in[0];
+  }
 
   // traverse
   CF_MDS_RetryRequestFactory cf(mdcache, mdr, true);
@@ -4284,6 +4289,11 @@ void Server::handle_client_open(MDRequestRef& mdr)
   if (!cur)
     return;
 
+  MutationImpl::LockOpVec lov;
+  lov.add_rdlock(&dnl->get_inode()->snaplock);
+  if (!mds->locker->acquire_locks(mdr, lov))
+    return;
+
+  MutationImpl::LockOpVec lov;
+  lov.add_rdlock(&dnl->get_inode()->snaplock);
+  if (!mds->locker->acquire_locks(mdr, lov))
+    return;
+
   if (cur->is_frozen() || cur->state_test(CInode::STATE_EXPORTINGCAPS)) {
     ceph_assert(!need_auth);
     mdr->locking_state &= ~(MutationImpl::PATH_LOCKED | MutationImpl::ALL_LOCKED);
@@ -4505,13 +4515,7 @@ void Server::handle_client_openc(MDRequestRef& mdr)
 
   CDentry::linkage_t *dnl = dn->get_projected_linkage();
   if (!excl && !dnl->is_null()) {
-    // it existed.
-    mds->locker->xlock_downgrade(&dn->lock, mdr.get());
-
-    MutationImpl::LockOpVec lov;
-    lov.add_rdlock(&dnl->get_inode()->snaplock);
-    if (!mds->locker->acquire_locks(mdr, lov))
-      return;
+    mds->locker->drop_locks(mdr.get());
 
     handle_client_open(mdr);
     return;
```

Reply:
> (1)curdir(cidr3)'s auth mds (2)testdn's dentry's auth mds (3)testdn's inode's auth mds
> 
> I think curdir(cidr3)'s auth mds is the same with testdn's dentry's auth mds. 1, Is this right? 2, If a file is hardlink, in its dentry, can we check it is a hardlink easily ?

Yeah, but the (1)(2) and (3) are different.

Reply:
So the question can be translated:
In path_traverse():
>want_dentry is true: this request should be on testdn's dentry's auth mds.
>want_auth is true: this request should be on testdn's inode's auth mds.

BUT:
For the hardlink file, testdn's dentry's auth mds is different with testdn's inode's auth mds.
In handle_client_openc->rdlock_path_xlock_dentry->path_traverse, MDS_TRAVERSE_WANT_DENTRY and MDS_TRAVERSE_WANT_AUTH are both setted, so logically the path_traverse() will can not forward to correst mds. This looks like a logiccal error .

Reply:
The impact of this problem:
1,
rdlock_path_xlock_dentry set MDS_TRAVERSE_WANT_DENTRY and MDS_TRAVERSE_WANT_DENTRY simultaneously.
handle_client_openc
handle_client_mknod
handle_client_mkdir
handle_client_symlink
handle_client_link
handle_client_unlink
This functions use rdlock_path_xlock_dentry. In some function, the mdr->get_filepath() can be a hardlink file path.
If the hardlink file path(dentry)'s auth mds is different with the hardlink file's inode's auth mds, the dn (which is from rdlock_path_xlock_dentry's return) will be not auth. In this functions, they may has assumed that the dn is on auth; If they cannot deal with the case dn is not auth well, it will make some error, just link this pr's crash.

2,
rdlock_two_paths_xlock_destdn also set MDS_TRAVERSE_WANT_DENTRY and MDS_TRAVERSE_WANT_DENTRY simultaneously, and it is called by handle_client_rename and handle_client_link. 
So, now handle_client_rename and handle_client_link have the same risk.

Reply:
> So the question can be translated: In path_traverse(): want_dentry is true: this request should be on testdn's dentry's auth mds. want_auth is true: this request should be on testdn's inode's auth mds.
> 

The following comments about this:

```
 811    * MDS_TRAVERSE_WANT_DENTRY: Caller wants tail dentry. Add a null dentry if
 812    * tail dentry does not exist. return 0 even tail dentry is null.
 813    * MDS_TRAVERSE_WANT_AUTH: Always forward request to auth MDS of target inode                                                                   
 814    * or auth MDS of tail dentry (MDS_TRAVERSE_WANT_DENTRY is set).
```
But it seems the comments doesn't match the code, the code will forward the request for tail dentry if 
 the MDS_TRAVERSE_WANT_DENTRY is NOT set.

Reply:
> > So the question can be translated: In path_traverse(): want_dentry is true: this request should be on testdn's dentry's auth mds. want_auth is true: this request should be on testdn's inode's auth mds.
> 
> The following comments about this:
> 
> ```
>  811    * MDS_TRAVERSE_WANT_DENTRY: Caller wants tail dentry. Add a null dentry if
>  812    * tail dentry does not exist. return 0 even tail dentry is null.
>  813    * MDS_TRAVERSE_WANT_AUTH: Always forward request to auth MDS of target inode                                                                   
>  814    * or auth MDS of tail dentry (MDS_TRAVERSE_WANT_DENTRY is set).
> ```
> 
> But it seems the comments doesn't match the code, the code will forward the request for tail dentry if the MDS_TRAVERSE_WANT_DENTRY is NOT set.

I doubt if that was intended, esp. when the "first" with/dentry check is
```
if (want_auth && want_dentry && depth == path.depth() - 1) {
...
}
```

So, why check if want_dentry is unset later.

Reply:
Yeah, from my understanding it's intended.

```
if (want_auth && want_dentry && depth == path.depth() - 1) {
...
}
```
This is for tail dentry's parent `CDir` and will make sure that the tail dentry's parent CDir is auth in current MDS, because it need to create a null dentry if the tail dentry does not exist, etc.

```
if (want_auth && !want_dentry) {
...
}
```
For this because if the `want_dentry` is true it usually means it only want to get the dentry itself and do not care about the corresponding `CInode` is auth or not, such as from `rdlock_path_xlock_dentry` caller case.

Reply:
Great analysis！
So, logiccally when rdlock_path_xlock_dentry return a not null dentry, this mds will be the dentry's auth.
But, if the file is hardlink, the inode's auth may be not this mds, so in handle_client_openc->handle_client_open case, handle_client_open with CEPH_O_TRUNC flag cannot use the inode directly.

Reply:
> Yeah, from my understanding it's intended.
> 
> ```
> if (want_auth && want_dentry && depth == path.depth() - 1) {
> ...
> }
> ```
> 
> This is for tail dentry's parent `CDir` and will make sure that the tail dentry's parent CDir is auth in current MDS, because it need to create a null dentry if the tail dentry does not exist, etc.
> 
> ```
> if (want_auth && !want_dentry) {
> ...
> }
> ```
> 
> For this because if the `want_dentry` is true it usually means it only want to get the dentry itself and do not care about the corresponding `CInode` is auth or not, such as from `rdlock_path_xlock_dentry` caller case.

Hmmm... the comment is a bit out-of-sync then. And the problem of using a non-auth inode with O_TRUNC exists...

Reply:
Yeah, it needs to be fixed anyway.

Reply:
Maybe the caller doesn't need the tail dentry to be auth if it can get an auth target inode.